### ISSUE TITLE
Helm upgrade failed for release, <HELM CHART> has no deployed releases

### ISSUE THREAD

**Issue by Nathan-Nesbitt** [fluxcd/flux2#4614]
### Describe the bug

Here is the config file that I have for creating the `rook-ceph-cluster` as defined [here]().

```yaml
apiVersion: helm.toolkit.fluxcd.io/v2beta1
kind: HelmRelease
metadata:
  name: rook-ceph-cluster
  namespace: rook-ceph
spec:
  chart:
    spec:
      chart: rook-ceph-cluster
      version: 1.13.x
      sourceRef:
        kind: HelmRepository
        name: rook
        namespace: flux-system
  interval: 30m
  timeout: 10m
  install:
    remediation:
      retries: 3
  upgrade:
    remediation:
      retries: -1 
    crds: CreateReplace 
  releaseName: rook-ceph-cluster
  values:
    # ALL THE VALUES FROM values.yaml
```

When I go to deploy, I get the following error, which I cannot for the life of me figure out why. 

```
Helm upgrade failed for release rook-ceph/rook-ceph-cluster with chart rook-ceph-cluster@v1.13.4: "rook-ceph-cluster" has no deployed releases
```

I can create it manually from the cli using helm by running the following:

```sh
helm repo add rook https://charts.rook.io/release
helm install my-rook-ceph-cluster rook/rook-ceph-cluster --version 1.13.4
```

So it's not an issue with the helm repository, or with rook.

Any ideas?

### Steps to reproduce

1. Create a K8s cluster
2. Create a setup based on the example git repo
3. Try to install rook-ceph using the above config (1/2 files, there is a lot of setup for this, but the other files do not cause an error)

### Expected behavior

It should just run as expected 🤷 

### OS / Distro

Ubuntu -- Jammy

### Flux version

flux version 2.2.3

### Flux check

► checking prerequisites
✔ Kubernetes 1.27.10+rke2r1 >=1.26.0-0
► checking version in cluster
✔ distribution: flux-v2.2.3
✔ bootstrapped: true
► checking controllers
✔ helm-controller: deployment ready
► ghcr.io/fluxcd/helm-controller:v0.37.4
✔ kustomize-controller: deployment ready
► ghcr.io/fluxcd/kustomize-controller:v1.2.2
✔ notification-controller: deployment ready
► ghcr.io/fluxcd/notification-controller:v1.2.4
✔ source-controller: deployment ready
► ghcr.io/fluxcd/source-controller:v1.2.4
► checking crds
✔ alerts.notification.toolkit.fluxcd.io/v1beta3
✔ buckets.source.toolkit.fluxcd.io/v1beta2
✔ gitrepositories.source.toolkit.fluxcd.io/v1
✔ helmcharts.source.toolkit.fluxcd.io/v1beta2
✔ helmreleases.helm.toolkit.fluxcd.io/v2beta2
✔ helmrepositories.source.toolkit.fluxcd.io/v1beta2
✔ kustomizations.kustomize.toolkit.fluxcd.io/v1
✔ ocirepositories.source.toolkit.fluxcd.io/v1beta2
✔ providers.notification.toolkit.fluxcd.io/v1beta3
✔ receivers.notification.toolkit.fluxcd.io/v1
✔ all checks passed

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by Nathan-Nesbitt** [comment]
The only thing I found was I was having similar issues with another helm repo, and the only way I could fix it was to nuke the entire flux bootstrap and restart. Saying that, that's not really practical for anything beyond setup... is there some other way to force reset helm (if that is actually the issue?)

**Comment by 98jan** [comment]
Hi,

what I can see is that the HelmRepository manifest is named "roock-ceph", but in the HelmRelease the referenced HelmRepository name ist "rook".

What helps me often is to run following command:
`kubectl describe HelmRelease rook-ceph-cluster -n rook-ceph`

**Comment by Nathan-Nesbitt** [comment]
@98jan thanks for the reply, I must have changed that last night while playing around, unfortunately it's still stuck with the error.

**Comment by Nathan-Nesbitt** [comment]
The interesting part to me is that, the error has to do with releases not the chart itself, which to me looks like the release is the issue as it finds `v1.13.14` in the helm repository which is not specified.

To me this looks like the release name is somehow wrong/non-existent? Even tho it is specified and should be the same as the name of the HelmRelease from above?

**Comment by Nathan-Nesbitt** [comment]
Yeah there's something very unusual going on, nuked the node and recreated everything and it's now working:

```
True    Helm install succeeded for release rook-ceph/rook-ceph-cluster.v1 with chart rook-ceph-cluster@v1.13.4
```
No change to the charts or anything, just started working after I deleted everything...

Is there any general advice on how to debug issues like this? Seems like general flakeyness but it's not clear why this worked or even what the actual original issue was.

**Comment by anakineo** [comment]
Not exactly the same situation but I had the same error. The helm release was previously installed by `Helm Install`.  I was porting it to use Flux to manage.

The fix was to remove the helm release secret.

I tried to look into Helm-controller codebase, and the path I found that could result the error is when flux doing helm upgrade

1.  `Reconcile` of `HelmReleaseReconciler` calls [AtomicRelease.Reconcile](https://github.com/fluxcd/helm-controller/blob/d2371a2e4676d2d93f96620131a998f0f09cd4da/internal/controller/helmrelease_controller.go#L389)
2. `AtomicRelease.Reconcile]` tries to [determine helm release state](https://github.com/fluxcd/helm-controller/blob/d2371a2e4676d2d93f96620131a998f0f09cd4da/internal/reconcile/atomic_release.go#L184). 
3. Returned `state` is then used to determine the [next action](https://github.com/fluxcd/helm-controller/blob/d2371a2e4676d2d93f96620131a998f0f09cd4da/internal/reconcile/atomic_release.go#L192)
4. In my case, `unmanaged` state [returns an upgrade action](https://github.com/fluxcd/helm-controller/blob/main/internal/reconcile/atomic_release.go#L356-L362)
5. `upgrade.reconcile` is called as the next action  where it uses [`action.Upgrade`](https://github.com/fluxcd/helm-controller/blob/d2371a2e4676d2d93f96620131a998f0f09cd4da/internal/reconcile/upgrade.go#L83).
6. `action.Upgrade` is basically a wrapper around helm upgrade
```go
return upgrade.RunWithContext(ctx, release.ShortenName(obj.GetReleaseName()), chrt, vals.AsMap())
```
7. helm upgrade then does a [pre-upgrade check](https://github.com/helm/helm/blob/e5e3fac3ca06970ad52012fe8cd92b0d2761bff0/pkg/action/upgrade.go#L159) on the release
```go
	// finds the last non-deleted release with the given name
	lastRelease, err := u.cfg.Releases.Last(name)
	if err != nil {
		// to keep existing behavior of returning the "%q has no deployed releases" error when an existing release does not exist
		if errors.Is(err, driver.ErrReleaseNotFound) {
			return nil, nil, driver.NewErrNoDeployedReleases(name)
		}
		return nil, nil, err
	}
```
if the release name is not found then "has no deployed releases" error is returned.

In summary, if flux determines a release needs to perform upgrade but the release name for some reason doesn't exist, this error is returned. 

I don't how `helm-controller` reached that particular state but deleting the helm release secret would have flux do an install instead of upgrade. that solved my issue.

**Comment by anakineo** [comment]
Not exactly the same situation. I had the same error while porting release that was installed by `Helm Install` to use flux to mange.

The fix was to delete helm release secret

I tried to look into helm-controller code base, the path I found that could lead to this error is when `helm-controller` does upgrade but the releaseName doesnt exist

1. for every recocile, helm-controller tries to [determine the state](https://github.com/fluxcd/helm-controller/blob/d2371a2e4676d2d93f96620131a998f0f09cd4da/internal/reconcile/atomic_release.go#L184) of the current release then calculate the [next action for that state](https://github.com/fluxcd/helm-controller/blob/d2371a2e4676d2d93f96620131a998f0f09cd4da/internal/reconcile/atomic_release.go#L192)
2. In my case, I think one of helm-controller's `ReleaseStatusOutOfSync` `ReleaseStatusUnmanaged`, `ReleaseStatusFailed`  state was returned. at least the time when helm-controller checked the status of helmrelase
3. based on the returned state, an[ `Upgrade` action was returned](https://github.com/fluxcd/helm-controller/blob/d2371a2e4676d2d93f96620131a998f0f09cd4da/internal/reconcile/atomic_release.go#L356)
```go
	case ReleaseStatusUnmanaged:
		log.Info(msgWithReason("release not managed by controller", state.Reason))

		// Clear the history as we can no longer rely on it.
		req.Object.Status.ClearHistory()

		return NewUpgrade(r.configFactory, r.eventRecorder), nil
```
4. Ugrade then goes on calling [action.Upgrade](https://github.com/fluxcd/helm-controller/blob/d2371a2e4676d2d93f96620131a998f0f09cd4da/internal/reconcile/upgrade.go#L83) which is an wrapper around helm upgrade action
```go
func Upgrade(ctx context.Context, config *helmaction.Configuration, obj *v2.HelmRelease, chrt *helmchart.Chart,
	vals helmchartutil.Values, opts ...UpgradeOption) (*helmrelease.Release, error) {
	upgrade := newUpgrade(config, obj, opts)

	policy, err := crdPolicyOrDefault(obj.GetUpgrade().CRDs)
	if err != nil {
		return nil, err
	}
	if err := applyCRDs(config, policy, chrt, setOriginVisitor(v2.GroupVersion.Group, obj.Namespace, obj.Name)); err != nil {
		return nil, fmt.Errorf("failed to apply CustomResourceDefinitions: %w", err)
	}

	return upgrade.RunWithContext(ctx, release.ShortenName(obj.GetReleaseName()), chrt, vals.AsMap())
}
```
5. inside helm upgrade, it does a[ pre-upgrade check](https://github.com/helm/helm/blob/e5e3fac3ca06970ad52012fe8cd92b0d2761bff0/pkg/action/upgrade.go#L159) where if no realse with the name found, "has no deployed releases" error is returned
```go
	// finds the last non-deleted release with the given name
	lastRelease, err := u.cfg.Releases.Last(name)
	if err != nil {
		// to keep existing behavior of returning the "%q has no deployed releases" error when an existing release does not exist
		if errors.Is(err, driver.ErrReleaseNotFound) {
			return nil, nil, driver.NewErrNoDeployedReleases(name)
		}
		return nil, nil, err
	}
```
This seems unlikely to happen as helm-controller did check the state of the helmrelease when determing the state. I don't know helm package enough to tell  but I guess helm state was changed in between state was determined by flux and helm check the state to perform upgrade so that the release no longer exist

Deleting the helm secret worked  as helm-controller would determine the release not installed and performs an install instead of upgrade.

**Comment by mark-oconnor-fidelity** [comment]
I encountered this problem today. The flux "HelmRelease" resource was failing to install. 

```
$ kubectl get hr velero
NAME     AGE     READY   STATUS
velero   2m58s   False   Helm upgrade failed for release cluster-addons/cluster-addons-velero with chart velero@6.0.0: "cluster-addons-velero" has no deployed releases
```

What I couldn't understand was why the HelmRelease was performing an upgrade action, when the ```helm ls``` command reported no releases:

```
$ helm history cluster-addons-velero
```

The root cause (in my case) was that there had been a previous installation of velero, which had not been properly cleaned up. 

```
$ helm history cluster-addons-velero
REVISION        UPDATED                         STATUS          CHART           APP VERSION     DESCRIPTION                              
1               Fri Jun  7 10:13:05 2024        uninstalling    velero-6.6.0  1.13.2          Deletion in progress (or silently failed)
```

**Comment by berlincount** [comment]
try ´kubectl get secret` in the respective namespace where you are managing everything. nuke any `sh.helm.release.v1.whatever` secret that's no actually applying. it'll try an install instead of an upgrade.

good luck!

**Comment by siddjellali** [comment]
Thanks

**Comment by Cajga** [comment]
Deleting the secret was the solution for us as well. Interestingly helm list did not show anything but the secret was indeed there... Possibly the content of it was not correct...

**Comment by gecube** [comment]
Triaged this, because the "nuke the cluster and it works again" outcome keeps biting
people. The error is fully reproducible, it is not environment flakiness, and it is
still present on current versions (flux v2.8.8 / helm-controller v1.5.5 / Helm v4, and
the same logic is in v0.37.x / Helm v3.13 that the reporter ran). Root cause is an
unhandled Helm release status: **`uninstalling`** — plus a second defect that makes
deleting and recreating the HelmRelease *not* recover from it, which is exactly why
re-bootstrapping didn't help here.

## TL;DR

A Helm uninstall that gets interrupted (controller pod restart/OOM mid-uninstall, an
API error / admission webhook denying a DELETE, a killed `helm uninstall`, ...) leaves
the release secret in Helm storage with `status: uninstalling` and description
`Deletion in progress (or silently failed)` — exactly what @mark-oconnor-fidelity
captured with `helm history`.

helm-controller has explicit handling for the *pending-\** statuses (unlock → mark
failed) and for `uninstalled` (treated as absent → install), but `uninstalling` falls
through every check and is classified as **Unmanaged**, for which the controller runs a
Helm **upgrade**. Helm's upgrade tolerates a missing deployed release only when the
last release is `failed` or `superseded`, so it returns `"<name>" has no deployed
releases`. The upgrade fails before anything is written to storage, so no failure
counter is ever incremented, remediation never kicks in, and the object retries this
forever. `flux reconcile hr <name> --force` goes through the same Unmanaged→upgrade
path, which is why forcing it only yields `✗ context deadline exceeded` (reproduced,
same output as in the issue description).

## Mechanism (code refs: helm-controller v1.5.5 / Helm v4.2.0; same in v0.37.4 / Helm v3.13.3)

1. Helm writes `uninstalling` into storage *before* deleting resources, and on a
   resource-deletion error returns *before* the final status update — the release
   stays `uninstalling` forever:
   https://github.com/helm/helm/blob/v3.13.3/pkg/action/uninstall.go#L103-L126
2. `DetermineReleaseState`: `uninstalling` is not `IsPending()` (no unlock), and only
   `uninstalled` maps to Absent → the release is declared Unmanaged ("found existing
   release in storage" when `status.history` is empty, "release not observed to be
   made for object" otherwise):
   https://github.com/fluxcd/helm-controller/blob/v1.5.5/internal/reconcile/state.go#L100-L112
3. Unmanaged → **clear history** → run upgrade, with no retries-exhausted check:
   https://github.com/fluxcd/helm-controller/blob/v1.5.5/internal/reconcile/atomic_release.go#L401-L407
4. Helm `prepareUpgrade` finds no deployed release; that is only tolerated for
   `failed`/`superseded` last releases:
   https://github.com/helm/helm/blob/v3.13.3/pkg/action/upgrade.go#L192-L222
   → `"<name>" has no deployed releases`.

## Why deleting/recreating the HelmRelease doesn't recover (second defect)

Step 3 above cleared `status.history` on the very first failed loop. When the
HelmRelease is later deleted, the uninstall reconciler requires a latest release in
history and returns `ErrNoLatest`, which the deletion path treats as
"nothing to do":
https://github.com/fluxcd/helm-controller/blob/v1.5.5/internal/reconcile/uninstall.go#L91-L93
https://github.com/fluxcd/helm-controller/blob/v1.5.5/internal/controller/helmrelease_controller.go#L557-L560

Verified live: `kubectl delete hr` succeeds (finalizer removed), but the poisoned
`sh.helm.release.v1.<name>.v1` secret **and all release resources stay in the
cluster**. A recreated HelmRelease (or a full flux re-bootstrap — @Nathan-Nesbitt's
case) immediately hits "found existing release in storage" → upgrade → the same error.
The cycle can only be broken by touching the secret, which is why nuking the node
"fixed" it.

## Reproduction (flux v2.8.8, kind, ConfigMap-only chart)

Any interruption works; the deterministic one is a webhook that denies DELETE:

```yaml
apiVersion: admissionregistration.k8s.io/v1
kind: ValidatingWebhookConfiguration
metadata: {name: deny-delete-4614}
webhooks:
  - name: deny-delete.example.com
    admissionReviewVersions: ["v1"]
    sideEffects: None
    failurePolicy: Fail
    clientConfig: {service: {name: nonexistent-webhook-svc, namespace: default, path: /validate}}
    rules: [{apiGroups: [""], apiVersions: ["v1"], operations: ["DELETE"], resources: ["configmaps"], scope: Namespaced}]
    namespaceSelector: {matchLabels: {kubernetes.io/metadata.name: repro4614}}
```

```console
$ helm uninstall demo4614 -n repro4614      # with the webhook in place
Error: failed to delete release: demo4614
$ helm history demo4614 -n repro4614
REVISION  UPDATED                   STATUS        CHART           DESCRIPTION
1         Thu Jul 16 10:29:18 2026  uninstalling  demo4614-0.1.0  Deletion in progress (or silently failed)
```

Remove the webhook, let the HelmRelease reconcile — helm-controller logs:

```
release not managed by controller: release not observed to be made for object
running 'upgrade' action with timeout of 2m0s
Reconciler error: "demo4614" has no deployed releases
release not managed by controller: found existing release in storage
running 'upgrade' action ...            # repeats forever; --force takes the same path
```

HelmRelease status is byte-for-byte the reported one:

```
Helm upgrade failed for release repro4614/demo4614 with chart demo4614@0.1.0+...: "demo4614" has no deployed releases
```

Side note on the `helm ls` confusion earlier in the thread (@Cajga): Helm 3's
`helm list` hides `uninstalling` releases by default — `helm list -a` / `helm history`
show them, and `kubectl get secret -l owner=helm --show-labels` shows
`status=uninstalling`. (Helm 4 lists them by default.)

## Recovery without nuking anything (all three verified live)

1. `helm uninstall <name> -n <ns>` (again) — Helm has no guard against re-running an
   `uninstalling` uninstall; it completes the deletion and purges the storage secret.
2. Or delete the stuck secret: `kubectl delete secret sh.helm.release.v1.<name>.v1
   -n <ns>` → next reconcile does a clean install (this is what worked for
   @berlincount / @Cajga; after it, the controller logged
   `Helm install succeeded for release ... .v1` — same as the reporter saw after
   recreating the node).
3. Deleting the HelmRelease object does **not** clean up (see second defect above).

## Proposed fix (happy to send a PR)

Treat `uninstalling` the way stale pending states are treated:

- in `DetermineReleaseState`, classify `common.StatusUninstalling` as
  `ReleaseStatusLocked`;
- in `Unlock`, extend `status.IsPending()` to also cover `StatusUninstalling`, marking
  the release `failed` ("Release unlocked from stale 'uninstalling' state").

Everything downstream then already works: the next reconcile sees a `failed` release
and Helm's upgrade explicitly tolerates a `failed` last release (upgrade.go link
above), recreating any half-deleted resources. No new states, no API changes,
symmetric with existing unlock semantics; regression tests in `state_test.go` /
`unlock_test.go`.

The deletion-path defect (silently skipping uninstall on `ErrNoLatest` while a release
for the object may still exist in the configured storage namespace) probably deserves
its own issue in helm-controller — can file that too.

Related: fluxcd/helm-controller#891 (same mechanism via manual `helm uninstall`),
fluxcd/helm-controller#367 (added the `uninstalled` → install path; `uninstalling` was
left out), fluxcd/helm-controller#804.
